DS13/code/modules/overmap/components.dm

```diff
@@ -0,0 +1,623 @@
+	if(supply_to)
+		supply_to.plasma_volume += powerproduction_drain
+	if(generator)
+		generator.plasma_volume += powerproduction_drain
+
+/obj/structure/overmap_component/plasma_injector/proc/find_supply_to()
+	for(var/obj/structure/overmap_component/integrity_field_generator/IFS in linked.powered_components)
+		if(!IFS.supplier)
+			generator = IFS
+			generator.supplier = src
+			supplying = TRUE
+	for(var/obj/structure/overmap_component/plasma_relay/X in linked.powered_components)
+		if(!X.supplier)
+			supply_to = X
+			X.supplier = src
+			supplying = TRUE
+			return TRUE
+
+/obj/structure/overmap_component/plasma_injector/proc/eject()
+	locked = FALSE
+	var/obj/item/tank/internals/plasma/Z = loaded_tank
+	if (!Z)
+		return
+	Z.forceMove(drop_location())
+	Z.layer = initial(Z.layer)
+	Z.plane = initial(Z.plane)
+	loaded_tank = null
+	icon_state = "injector"
+
+/obj/structure/overmap_component/plasma_relay //Annnd these supply power to your systems
+	name = "ODN relay"
+	desc = "This sturdy fuses warp plasma into power via the power grid. This power is then supplied to a subsystem of choice"
+	icon_state = "relay-closed"
+	req_access = list(ACCESS_ENGINE_EQUIP)
+	var/obj/structure/overmap_component/plasma_injector/supplier
+	var/supplying = null //can be engines, shields, weapons
+	var/panel_open = FALSE //You gotta take the panel off first if you want to mess with it.
+	var/plasma_volume = 0 //How much plasma we got? If you store loads in it and this takes a hit it'll rupture and spill some of it all over the place.
+	var/plasma_drain = 0.005 //Allows a tiny bit of plasma buildup in normal usage cases.
+	var/active = FALSE //Have we got enough plasma to power a system?
+	var/activated = FALSE //Have we powered a system? If so, no need to spam its max_power stat endlessly
+	var/list/zaps = list('DS13/sound/effects/damage/consolehit.ogg','DS13/sound/effects/damage/consolehit2.ogg','DS13/sound/effects/damage/consolehit3.ogg','DS13/sound/effects/damage/consolehit4.ogg')
+	var/list/bleeps = list('DS13/sound/effects/computer/bleep1.ogg','DS13/sound/effects/computer/bleep2.ogg')
+	var/repair_step = "wrench" //Wrench, screwdriver, wirecutters, screwdriver, wrench
+	max_integrity = 150
+	var/benefit = 2 //2 of these per subsystem, or build more thru RnD
+	var/force_shutdown = FALSE //Are we being forced to turn off?
+
+/obj/structure/overmap_component/plasma_relay/attack_hand(mob/user)
+	. = ..()
+	var/Q = alert(user,"Redirect power to what system?","[src]","shields","weapons", "engines")
+	if(!Q || Q == "cancel")
+		playsound(loc, 'DS13/sound/effects/computer/bleep2.ogg',100)
+		return
+	playsound(loc, 'DS13/sound/effects/computer/beep3.ogg',100)
+	set_active(FALSE) //Remove us from our benefactor
+	supplying = Q
+	set_active(TRUE)
+	to_chat(user, "<span class='notice'>[src] will now supply the [supplying] subsystem with power</span>")
+
+/obj/structure/overmap_component/plasma_relay/attackby(obj/item/I, mob/user, params)
```

Comment:
Use the tool_act procs instead.

Reply:
You got a list of them handy? I couldnt find them at all

Reply:
https://github.com/DDMers/Deep-Space-13/blob/6743365585ddb995b9336fe04da9c47c0c172a40/code/game/atoms.dm#L567

If you convert them to tool_acts, make them return TRUE, since the tool will hit the machine if it returns FALSE

Reply:
a lot of stuff uses screwdriver_act and wrench_act now look at that to work it out